### PR TITLE
Fix GetAsync  not work when option is AllVersion

### LINKED ISSUE #16027: Remove `VersionOptions.AllVersions`
For an explanation, see comments below from https://github.com/OrchardCMS/OrchardCore/issues/16027#issuecomment-2115840661. We should remove `VersionOptions.AllVersions`.

The original issue below is kept for context.

### Describe the bug

The method `_contentManager.GetAsync(model.ContentItemId, VersionOptions.AllVersions)` **fails to return any content items, whether published or draft.** When debugging in Visual Studio with "Just my code" disabled, execution enters the condition `else if (options.IsDraft || options.IsDraftRequired)` in `DefaultContentManager` [here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L189). 

I'm unsure why this condition is satisfied, given that my content item is published and the boolean `IsDraft` and `IsDraftRequired` are set to `false`. 

Screenshots are provided for reference.

### To Reproduce
1. Use Orchard Core 1.8.3
2. Activate the "Content Fields Indexing (SQL)" feature (I have it active)
3. Create a Content Item and publish it
4. In a controller, use `IContentManager` from dependency injection to retrieve the Content Item by its ID with `VersionOptions.AllVersions` set, like this: `_contentManager.GetAsync(model.ContentItemId, VersionOptions.AllVersions)`
5. Execute the code to verify whether a Content Item is returned

### Expected behavior
A published or drafted Content Item must be returned when I set `VersionOptions.AllVersions`

### Screenshots

My Content Item:

image

`IsDraft` and `IsDraftRequired` are false, why the condition is satisfied?

image

### PR DESCRIPTION
**Pull request by hyzx86**
Fixes #16027

### CODE REVIEW

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs [hunk 1]
```diff
@@ -171,44 +171,54 @@ public async Task<ContentItem> GetAsync(string contentItemId, VersionOptions opt
...
-            else if (options.IsDraft && !options.IsDraftRequired)
+            if (options.IsAllVersions)
             {
                 contentItem = await _session
                     .Query<ContentItem, ContentItemIndex>()
                     .Where(x =>
-                        x.ContentItemId == contentItemId &&
-                        x.Published == false &&
-                        x.Latest == true)
+                        x.ContentItemId == contentItemId)
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:179]: Hmm, I think better to not take into account soft deleted items having both Published = Latest = false.

    x.ContentItemId == contentItemId && (x.Latest || x.Published)

Comment (hyzx86) [reply src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:179]: Not really, if you select AllVersion but it is still restricted. It's confusing

Comment (hyzx86) [reply src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:179]: > Hmm, I think better to not take into account soft deleted items having both Published = Latest = false.

@jtkech Yes, you are right, if you want to get different versions, you need to specify the version number.

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: We need to have a DeletedVersions option then too I think.

Comment (hyzx86) [reply src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: But the current content state is a little bit too complicated. For example, as a new person, you need to look up code or documentation to fully understand what the state means

**Comment by sebastienros** [comment]
Can you describe what "doesn't work" so we can recommend the best approach? Maybe open an issue?

**Comment by Skrypt** [comment]
Retrieving AllVersions all at once with the ContentManager.

**Comment by jtkech** [comment]
Or as I understood, return any active version, maybe in priority the published version if it exists, or a draft version if it exists, that' why here the suggested code uses `contentItemId && (x.Latest || x.Published)`.

But currently it then uses `FirstOrDefaultAsync()`, so it may return a draft whereas a published version exists.

Maybe it should use `ListAsync()` allowing to return a published version in priority.

**Comment by hyzx86** [comment]
Yes, it is the same problem as this one, 
https://github.com/OrchardCMS/OrchardCore/pull/11447 

Because my program uses SPA mode, much of the data manipulation is done through the Api
Initially I thought the draft state would be saved as the new ContentEntitemId,
It turns out there may be something wrong with the data I'm transferring, so the new ContentItemId is generated because the POST API/content doesn't seem to have any validation.
It might need to exclude two status fields (latest, published)(I can't remember),
For simplicity, I've implemented an API similar to the Graphql mutation myself
It can directly convert Graphql query results to Contentitem and then update the content

https://github.com/EasyOC/EasyOC/blob/eb1dc57f37491296823b633c30dc30806613e573/src/Modules/EasyOC.OrchardCore.ContentExtentions/AppServices/ContentManagementAppService.cs#L67

**Comment by hyzx86** [comment]
> Can you describe what "doesn't work" so we can recommend the best approach? Maybe open an issue?

mean is this filter option is never used in this method, But now it seems that just changing this doesn't make much sense, because if no filtering conditions are attached, a ContentItem will be obtained randomly (or with a minimum ID), so at least one ContentVersionId needs to be specified here

image

**Comment by hyzx86** [comment]
Since this option doesn't serve any purpose, can we remove it?
It only has one reference in the Graphql status filter

https://github.com/OrchardCMS/OrchardCore/blob/5d3cdca6cd2e04a404f130ca38a404a2699f37dc/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/IContentManager.cs#L265

**Comment by jtkech** [comment]
Maybe we could have used AllVersions for the GetAsync() that returns a collection, currently we use a `bool latest = false` parameter, false meaning return all published, true all latest regardless there are published or not. So yes here we could have used AlVersions but we would need a new method to not break the existing one.

For the GetAsync that returns only one item, hmm maybe we could have used an AnyVersion with the behavior I described here https://github.com/OrchardCMS/OrchardCore/pull/11433#issuecomment-1082390420, but was just an idea, maybe too confusing ;)

So yes, if we keep the existing code I agree with you, maybe better to just get rid of the AllVersions option, or as said above and if we think it is worth, maybe use it for a GetAsync that returns a collection of items.

Note: Here we are talking about active versions (Published or Latest), Published and Latest are not versions in themselves but properties of a given version (related to a given contentItemVersionId), a content item having some versions but no active versions is a soft removed item.

**Comment by hyzx86** [comment]
https://github.com/OrchardCMS/OrchardCore/pull/11447

**Comment by hyzx86** [comment]
@jtkech 
This is a very embarrassing question for me, if I want to rewrite `GetAsync(id,option)`  ,  I need call `BuildNewVersionAsync`
But since it is not public, I have to copy this method to my  `ContentManager` 

https://github.com/OrchardCMS/OrchardCore/blob/4093d2e71fc9ef48c96afd609a6564826e17b803/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L239

But even if I copied it, I found `ContentElement.Data` is not open either

https://github.com/OrchardCMS/OrchardCore/blob/4093d2e71fc9ef48c96afd609a6564826e17b803/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L416

Finally, I wrote l index query directly by referring to the code in AdminController
If so, what is the meaning of content manager

https://github.com/OrchardCMS/OrchardCore/blob/4093d2e71fc9ef48c96afd609a6564826e17b803/src/OrchardCore.Modules/OrchardCore.Contents/Controllers/AdminController.cs#L310

**Comment by jtkech** [comment]
> Finally, I wrote l index query directly by referring to the code in AdminController
If so, what is the meaning of ContentManager

It allows to call the contentManager LoadAsync() on each contentItem which call handlers to activate correctly items if not yet done by looking at the contentManagerSession cache.

**Comment by hyzx86** [comment]
Yes, I see. I mean, the current implementation of content manager is not complete enough

**Comment by jtkech** [comment]
As additional info, as it is done when using content manager GetAsync(), activating item meaning activating item parts/fields

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:1]: Please provide unit tests.

**Review by Piedone — CHANGES_REQUESTED** [review]

**Comment by Piedone** [comment]
We should rather remove `VersionOptions.AllVersions`, see comments from VersionOptions.AllVersions.

**Pull request closed without merging**